src/libutil/logging.cc

```diff
@@ -289,9 +305,49 @@ struct JSONLogger : Logger {
     }
 };
 
-std::unique_ptr<Logger> makeJSONLogger(Descriptor fd)
+std::unique_ptr<Logger> makeJSONLogger(Descriptor fd, bool includeNixPrefix)
+{
+    return std::make_unique<JSONLogger>(fd, includeNixPrefix);
+}
+
+std::unique_ptr<Logger> makeJSONLogger(const std::filesystem::path & path, bool includeNixPrefix)
+{
+    struct JSONFileLogger : JSONLogger {
+        AutoCloseFD fd;
+
+        JSONFileLogger(AutoCloseFD && fd, bool includeNixPrefix)
+            : JSONLogger(fd.get(), includeNixPrefix)
+            , fd(std::move(fd))
+        { }
+    };
+
+    AutoCloseFD fd =
+        std::filesystem::is_socket(path)
+        ? connect(path)
+        : toDescriptor(open(path.c_str(), O_CREAT | O_APPEND | O_WRONLY, 0644));
```

Comment:
This will allow multiple Nix instances to append to the same file simultaneously, but the writing isn't synchronized between them, which will lead to interleaved (corrupted) entries, especially if some of the log lines are larger than the page size (e.g. long `gcc` invocation logs from `make`, but could be anything)

Reply:
I think this should be not an issue for unix sockets because we can tell apart the connections? Maybe we should only support unix sockets than for the time being.

Reply:
One could still allow file log output by locking the file using an flock or similar, right(?), this is at least useful for debugging

Reply:
I was thinking of locking per-invocation since within the process there is a mutex lock for log writing.

Reply:
So we had a discussion in the nix meeting and for now it would be fine to document the concurrency issue in case we are using files and point to the unix socket as primary to consume this information.